### PR TITLE
Improve template grub2 argument

### PR DESCRIPTION
**Pull request by vojtapolasek**
#### Description:

The templates for checking and remediating of Grub2 arguments are improved. The way of updating Grub configuration between RHEL7 and RHEL8 has changed and therefore two ways of accomplishing checking and remediation were used. The documentation and OCIL was also updated.

#### Rationale:

The work started because of BZ [https://bugzilla.redhat.com/show_bug.cgi?id=1547642](https://bugzilla.redhat.com/show_bug.cgi?id=1547642), then it was found that the reworking because of changes between RH7 and RH8 is needed.

### CODE REVIEW

File: shared/templates/template_ANSIBLE_grub2_bootloader_argument [hunk 1]
```diff
@@ -25,3 +26,10 @@
 
 - name: update bootloader menu
   command: /sbin/grubby --update-kernel=ALL --args="{{{ ARG_NAME_VALUE }}}"
+
+{{% else %}}
+
+- name: update bootloader menu
+  shell: '/usr/bin/grub2-editenv - set "$(grub2-editenv - list | grep kernelopts) {{{ ARG_NAME_VALUE }}}"'
```
Comment (redhatrises) [inline shared/templates/template_ANSIBLE_grub2_bootloader_argument]: Please don't use shell and come up with 2 tasks to handle this.

File: linux_os/guide/system/auditing/grub2_audit_argument/rule.yml [hunk 2]
```diff
@@ -42,16 +48,40 @@ references:
 ocil_clause: 'auditing is not enabled at boot time'
 
 ocil: |-
+{{% if product == "rhel7" %}}
     Inspect the form of default GRUB 2 command line for the Linux operating system
     in <tt>/etc/default/grub</tt>. If they include <tt>audit=1</tt>, then auditing
     is enabled at boot time.
+    First check if the Grub recovery is enabled:
+    <pre>$ grep 'GRUB_DISABLE_RECOVERY' /etc/default/grub</pre>
+    If this option is set to true, then check that the line produced by
```
Comment (yuumasato) [inline linux_os/guide/system/auditing/grub2_audit_argument/rule.yml]: `check that a line is output by following command`

File: linux_os/guide/system/auditing/grub2_audit_argument/rule.yml [hunk 3]
```diff
@@ -42,16 +48,40 @@ references:
...
+    <pre>$ grep 'GRUB_CMDLINE_LINUX_DEFAULT.*audit=1.*' /etc/default/grub</pre>
+    exists. If the recovery is disabled, check the line with
```
Comment (yuumasato) [inline linux_os/guide/system/auditing/grub2_audit_argument/rule.yml]: For readability, break the line before talking about case when recovery is disabled.

File: linux_os/guide/system/auditing/grub2_audit_argument/rule.yml [hunk 4]
```diff
@@ -63,10 +93,10 @@ warnings:
         <li>On BIOS-based machines, issue the following command as <tt>root</tt>:
         <pre>~]# grub2-mkconfig -o /boot/grub2/grub.cfg</pre></li>
         <li>On UEFI-based machines, issue the following command as <tt>root</tt>:
-{{% if product in ["rhel7", "rhel8", "ol7", "ol8"] %}}
+{{% if product in ["rhel7", "ol7", "ol8"] %}}
```
[the remark this answers is not in the document]
Comment (vojtapolasek) [reply linux_os/guide/system/auditing/grub2_audit_argument/rule.yml]: @iokomin  according to documentation, setting kernel parameters for all kernels through GRUB2 are different for RH7 and RH8.
RH7: use grubby or grub2-mkconfig
rh8: use grub2-editenv
Is this the same for OL7 / OL8?
Thanks.

[the remark this answers is not in the document]
Comment (iokomin) [reply linux_os/guide/system/auditing/grub2_audit_argument/rule.yml]: @vojtapolasek in terms of GRUB2 support on OL it should follow RHEL guidelines, thus rhel7/rhel8 changes applicable to ol7/ol8 accordingly. Thus rhel7 specific conditional changes applicable to ol7 as well (exception are possible links on guides).
Note: agree with @yuumasato, this code line is unreachable for ol7 and ol8 taking into account condition on the line 84.

Thanks,
Ilya.

File: shared/templates/template_ANSIBLE_grub2_bootloader_argument [hunk 5]
```diff
@@ -29,7 +29,14 @@
 
 {{% else %}}
 
-- name: update bootloader menu
-  shell: '/usr/bin/grub2-editenv - set "$(grub2-editenv - list | grep kernelopts) {{{ ARG_NAME_VALUE }}}"'
+- name: get current kernel parameters
+  command: 'grep kernelopts `/usr/bin/grub2-editenv - list`'
```
Comment (redhatrises) [inline shared/templates/template_ANSIBLE_grub2_bootloader_argument]: This should be `command: /usr/bin/grub2-editenv list`

File: shared/templates/template_ANSIBLE_grub2_bootloader_argument [hunk 6]
```diff
@@ -29,7 +29,14 @@
...
+  register: kernelopts
```
Comment (redhatrises) [inline shared/templates/template_ANSIBLE_grub2_bootloader_argument]: Add `changed_when: False` after `register`

File: shared/templates/template_ANSIBLE_grub2_bootloader_argument [hunk 7]
```diff
@@ -29,7 +29,14 @@
...
+
+- set_fact:
+  kerneloptsfact: "{{ kernelopts.stdout }}"
```
Comment (redhatrises) [inline shared/templates/template_ANSIBLE_grub2_bootloader_argument]: Remove this `set_fact`

File: shared/templates/template_ANSIBLE_grub2_bootloader_argument [hunk 8]
```diff
@@ -29,7 +29,14 @@
...
+
+- name: Update the bootloader menu
+  command: '/usr/bin/grub2-editenv - set {{ kerneloptsfact }} {{{ ARG_NAME_VALUE }}}'
```
Comment (redhatrises) [inline shared/templates/template_ANSIBLE_grub2_bootloader_argument]: `command: /usr/bin/grub2-editenv set {{ item }} {{{ ARG_NAME_VALUE }}}`
 
Also need to add:
````
with_items: "{{ kernelopts.stdout_lines | select('match', '^kernelopts.*') | list }}"
when:
  - kernelopts.stdout_lines is defined
  - kernelopts.stdout_lines | length > 0
````

Comment (vojtapolasek) [reply shared/templates/template_ANSIBLE_grub2_bootloader_argument]: Thank you @redhatrises I am not very familiar with Ansible so this was a nice introduction.

**Comment by yuumasato** [comment]
@vojtapolasek Can you rebase into latest master?
Although the test scenarios are in the new place, within the rule folder, the Test Suite of the branch jis still looking for them in `tests/data`.

File: shared/templates/template_ANSIBLE_grub2_bootloader_argument (hunk 8 above)
Comment (redhatrises) [reply shared/templates/template_ANSIBLE_grub2_bootloader_argument]: @vojtapolasek yeah no problem. One question about `grub2-editenv` that I have is the dash supposed to exist like `/usr/bin/grub2-editenv - set` or is it supposed to be `/usr/bin/grub2-editenv set` without the dash?

Comment (vojtapolasek) [reply shared/templates/template_ANSIBLE_grub2_bootloader_argument]: The dash needst to be there, without it it asks for the target file location and the dash is also used in the rhel8 documentation.

**Comment by jan-cerny** [comment]
I have checked that the test scenraios are passing on both RHEL 8 and 7 for Bash remediations.

```
[jcerny@thinkpad scap-security-guide{improve_template_grub2_argument}]$ test_rule grub2_audit_argument rhel8
~/scap-security-guide/tests ~/scap-security-guide
Setting console output to log level INFO
INFO - The base image option has not been specified, choosing libvirt-based test environment.
INFO - Logging into /home/jcerny/scap-security-guide/tests/logs/rule-custom-2019-09-10-0912/test_suite.log
INFO - xccdf_org.ssgproject.content_rule_grub2_audit_argument
WARNING - Script wrong_value_rhel7.fail.sh is not applicable on given platform
WARNING - Script correct_recovery_disabled.pass.sh is not applicable on given platform
WARNING - Script arg_not_there_etcdefaultgrub.fail.sh is not applicable on given platform
WARNING - Script arg_not_there_rhel7.fail.sh is not applicable on given platform
WARNING - Script wrong_value_etcdefaultgrub.fail.sh is not applicable on given platform
WARNING - Script correct_grubby.pass.sh is not applicable on given platform
WARNING - Script arg_not_there_etcdefaultgrub_recovery_disabled.fail.sh is not applicable on given platform
WARNING - Script wrong_value_etcdefaultgrub_recovery_disabled.fail.sh is not applicable on given platform
INFO - Script arg_not_there_rhel8.fail.sh using profile xccdf_org.ssgproject.content_profile_ospp OK
INFO - Script correct_value.pass.sh using profile xccdf_org.ssgproject.content_profile_ospp OK
INFO - Script wrong_value_rhel8.fail.sh using profile xccdf_org.ssgproject.content_profile_ospp OK
INFO - Script correct_grubenv.pass.sh using profile xccdf_org.ssgproject.content_profile_ospp OK
```

```
[jcerny@thinkpad scap-security-guide{improve_template_grub2_argument}]$ test_rule grub2_audit_argument rhel7
~/scap-security-guide/tests ~/scap-security-guide
Setting console output to log level INFO
INFO - The DataStream contains 2 Benchmarks
INFO - 0 - scap_org.open-scap_cref_ssg-rhel7-xccdf-1.2.xml
INFO - 1 - scap_org.open-scap_cref_ssg-rhel7-pcidss-xccdf-1.2.xml
INFO - Selected Benchmark is 0
INFO - To select a different Benchmark, use --xccdf-id-number option.
INFO - The base image option has not been specified, choosing libvirt-based test environment.
INFO - Logging into /home/jcerny/scap-security-guide/tests/logs/rule-custom-2019-09-10-0918/test_suite.log
INFO - xccdf_org.ssgproject.content_rule_grub2_audit_argument
WARNING - Script arg_not_there_rhel8.fail.sh is not applicable on given platform
WARNING - Script wrong_value_rhel8.fail.sh is not applicable on given platform
WARNING - Script correct_grubenv.pass.sh is not applicable on given platform
INFO - Script wrong_value_rhel7.fail.sh using profile xccdf_org.ssgproject.content_profile_ospp OK
INFO - Script correct_recovery_disabled.pass.sh using profile xccdf_org.ssgproject.content_profile_ospp OK
INFO - Script arg_not_there_etcdefaultgrub.fail.sh using profile xccdf_org.ssgproject.content_profile_ospp OK
INFO - Script correct_value.pass.sh using profile xccdf_org.ssgproject.content_profile_ospp OK
INFO - Script arg_not_there_rhel7.fail.sh using profile xccdf_org.ssgproject.content_profile_ospp OK
INFO - Script wrong_value_etcdefaultgrub.fail.sh using profile xccdf_org.ssgproject.content_profile_ospp OK
INFO - Script correct_grubby.pass.sh using profile xccdf_org.ssgproject.content_profile_ospp OK
INFO - Script arg_not_there_etcdefaultgrub_recovery_disabled.fail.sh using profile xccdf_org.ssgproject.content_profile_ospp OK
INFO - Script wrong_value_etcdefaultgrub_recovery_disabled.fail.sh using profile xccdf_org.ssgproject.content_profile_ospp OK
```

**Comment by jan-cerny** [comment]
I have checked that the test scenraios are passing on both RHEL 8 and 7 for Ansible remediations.


```
[jcerny@thinkpad scap-security-guide{improve_template_grub2_argument}]$ test_rule grub2_audit_argument rhel8 ansible
~/scap-security-guide/tests ~/scap-security-guide
Setting console output to log level INFO
INFO - The base image option has not been specified, choosing libvirt-based test environment.
INFO - Logging into /home/jcerny/scap-security-guide/tests/logs/rule-custom-2019-09-10-1010/test_suite.log
INFO - xccdf_org.ssgproject.content_rule_grub2_audit_argument
WARNING - Script wrong_value_rhel7.fail.sh is not applicable on given platform
WARNING - Script correct_recovery_disabled.pass.sh is not applicable on given platform
WARNING - Script arg_not_there_etcdefaultgrub.fail.sh is not applicable on given platform
WARNING - Script arg_not_there_rhel7.fail.sh is not applicable on given platform
WARNING - Script wrong_value_etcdefaultgrub.fail.sh is not applicable on given platform
WARNING - Script correct_grubby.pass.sh is not applicable on given platform
WARNING - Script arg_not_there_etcdefaultgrub_recovery_disabled.fail.sh is not applicable on given platform
WARNING - Script wrong_value_etcdefaultgrub_recovery_disabled.fail.sh is not applicable on given platform
INFO - Script arg_not_there_rhel8.fail.sh using profile xccdf_org.ssgproject.content_profile_ospp OK
INFO - Script correct_value.pass.sh using profile xccdf_org.ssgproject.content_profile_ospp OK
INFO - Script wrong_value_rhel8.fail.sh using profile xccdf_org.ssgproject.content_profile_ospp OK
INFO - Script correct_grubenv.pass.sh using profile xccdf_org.ssgproject.content_profile_ospp OK
~/scap-security-guide
```

```
[jcerny@thinkpad scap-security-guide{improve_template_grub2_argument}]$ test_rule grub2_audit_argument rhel7 ansible
~/scap-security-guide/tests ~/scap-security-guide
Setting console output to log level INFO
INFO - The DataStream contains 2 Benchmarks
INFO - 0 - scap_org.open-scap_cref_ssg-rhel7-xccdf-1.2.xml
INFO - 1 - scap_org.open-scap_cref_ssg-rhel7-pcidss-xccdf-1.2.xml
INFO - Selected Benchmark is 0
INFO - To select a different Benchmark, use --xccdf-id-number option.
INFO - The base image option has not been specified, choosing libvirt-based test environment.
INFO - Logging into /home/jcerny/scap-security-guide/tests/logs/rule-custom-2019-09-10-0934/test_suite.log
INFO - xccdf_org.ssgproject.content_rule_grub2_audit_argument
WARNING - Script arg_not_there_rhel8.fail.sh is not applicable on given platform
WARNING - Script wrong_value_rhel8.fail.sh is not applicable on given platform
WARNING - Script correct_grubenv.pass.sh is not applicable on given platform
INFO - Script wrong_value_rhel7.fail.sh using profile xccdf_org.ssgproject.content_profile_ospp OK
INFO - Script correct_recovery_disabled.pass.sh using profile xccdf_org.ssgproject.content_profile_ospp OK
INFO - Script arg_not_there_etcdefaultgrub.fail.sh using profile xccdf_org.ssgproject.content_profile_ospp OK
INFO - Script correct_value.pass.sh using profile xccdf_org.ssgproject.content_profile_ospp OK
INFO - Script arg_not_there_rhel7.fail.sh using profile xccdf_org.ssgproject.content_profile_ospp OK
INFO - Script wrong_value_etcdefaultgrub.fail.sh using profile xccdf_org.ssgproject.content_profile_ospp OK
INFO - Script correct_grubby.pass.sh using profile xccdf_org.ssgproject.content_profile_ospp OK
INFO - Script arg_not_there_etcdefaultgrub_recovery_disabled.fail.sh using profile xccdf_org.ssgproject.content_profile_ospp OK
INFO - Script wrong_value_etcdefaultgrub_recovery_disabled.fail.sh using profile xccdf_org.ssgproject.content_profile_ospp OK
~/scap-security-guide
```

**Pull request merged**